File: py_payment_api/app/services/store/store_service.py
from typing import List, Dict, TypeVar, Generic, Optional, cast
from enum import StrEnum
from ..store.payment.abstract import ItemCollectionService
from .enum import AvailableItem, ItemType
from .payment import OneTimePaymentData, SubscriptionPaymentData, PurchaseDetail, PurchaseStatus
from typing import Any
ITEM_CATEGORY = TypeVar('ITEM_CATEGORY', bound=StrEnum)

class StoreServiceError(Exception):
    """Base exception for store service errors."""
    pass

class StoreService(Generic[ITEM_CATEGORY]):
    """Service for managing store items."""

    __item_collection_services: Dict[ItemType, ItemCollectionService[ITEM_CATEGORY]]

    def __init__(self, item_type_to_item_service : Dict[ItemType, ItemCollectionService[ITEM_CATEGORY]] | None = None) -> None:
        """Initialize the store service."""
        self.__item_collection_services = item_type_to_item_service or {}   

    def _get_service(self, item_type: ItemType) -> ItemCollectionService[ITEM_CATEGORY]:
        """Get the service for a specific item type."""
        service = self.__item_collection_services.get(item_type)
        if not service:
            raise StoreServiceError(f"No service available for item type: {item_type}")
        return service

    def get_item(self, item_type: ItemType, item_category: ITEM_CATEGORY, item_id: str) -> Optional[AvailableItem]:
        """Get a specific item."""
        service = self._get_service(item_type)
        result = service.get_item(item_category, item_id)
        return cast(Optional[AvailableItem], result)

    def get_all_items_by_type(self, item_type: ItemType) -> List[AvailableItem]:
        """Get all items of a specific type."""
        service = self._get_service(item_type)
        return cast(List[AvailableItem], service.get_all_items())
    
    def get_all_items(self) -> Dict[ItemType, List[AvailableItem]]:
        """Get all items grouped by type."""
        return {
            item_type: cast(List[AvailableItem], service.get_all_items())
            for item_type, service in self.__item_collection_services.items()
        }
    
    def get_orders_by_user_id(self, user_id: str) -> List[PurchaseDetail]:
        """Get all orders by user ID."""
        orders = []
        for service in self.__item_collection_services.values():
            orders.extend(service.get_orders_by_user_id(user_id))
        return orders

    def get_purchase_type(self, item_name: str) -> ItemType:
        """Get the purchase type of an item."""
        for service in self.__item_collection_services.values():
            if service.has_item(item_name):
                return service.item_type
        return ItemType.UNKNOWN
    
    def update_order_status(self, purchase_id: str, status: PurchaseStatus) -> None:
        """Update the status of an order."""
        updated = False
        for service in self.__item_collection_services.values():
            if service.has_order(purchase_id):
                service.change_order_status(purchase_id, status)
                updated = True
        if not updated:
            raise StoreServiceError(f"No order found with purchase ID: {purchase_id}")

File: tsPaymentApi/src/services/store/StoreService.ts
import { ItemType } from '../ItemType';
import { AvailableItem } from './AvailableItem';
import { PurchaseDetail } from './payment/PurchaseDetail';
import { PurchaseStatus } from './payment/PurchaseStatus';
import { ItemCollectionService } from './payment/abstract/ItemCollectionService';

export class StoreServiceError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'StoreServiceError';
    }
}

export class StoreService<ITEM_CATEGORY extends string> {
    private readonly itemTypeToItemService: Map<ItemType, ItemCollectionService<ITEM_CATEGORY>>;

    constructor(itemTypeToItemService: Map<ItemType, ItemCollectionService<ITEM_CATEGORY>>) {
        this.itemTypeToItemService = itemTypeToItemService;
    }

    public async getOrdersByUserId(userId: string): Promise<PurchaseDetail[]> {
        // TODO: Implement order retrieval from database
        return [];
    }

    public async getAllItemsByType(itemType: ItemType): Promise<AvailableItem[]> {
        const service = this.itemTypeToItemService.get(itemType);
        if (!service) {
            throw new StoreServiceError(`No service registered for item type: ${itemType}`);
        }
        return service.getItems();
    }

    public async getAllItems(): Promise<Record<ItemType, AvailableItem[]>> {
        const result: Record<ItemType, AvailableItem[]> = {} as Record<ItemType, AvailableItem[]>;
        for (const [type, service] of this.itemTypeToItemService.entries()) {
            result[type] = await service.getItems();
        }
        return result;
    }

    public async getPurchaseType(itemName: string): Promise<ItemType> {
        for (const [type, service] of this.itemTypeToItemService.entries()) {
            if (await service.validateItem(itemName)) {
                return type;
            }
        }
        throw new StoreServiceError(`No service found for item: ${itemName}`);
    }

    public async updateOrderStatus(orderId: string, status: PurchaseStatus): Promise<void> {
        // TODO: Implement order status update in database
    }
} 